### ISSUE TITLE
--jobs does not consider --local-cores when executing localrules

### ISSUE THREAD

**Issue by ckrushton** [snakemake/snakemake#2896]
**Snakemake version**

8.4.11

**Describe the bug**

I'm running a snakemake workflow on a SLURM cluster (using `snakemake-executor-plugin-cluster-generic`), using cluster nodes with 48 cores and a head node with 8 cores.

My profile file looks like this:
`jobs: 600`
`cores: 48`
`local-cores: 4`

My workflow has a rule which waits for all samples to complete, then executes an I/O intensive step for each sample. To minimize filesystem burden, I specify this rule as a localrule (due to filesystem sync issues between cluster nodes).

However, when Snakemake executes this step, it launches up to 600 jobs on the localnode instead of 4. I understand that, as `--jobs` is described, this makes sense ; however, it seems sensible that snakemake should consider the resources available on the local node when selecting how many local jobs to execute simultaneously.
**Logs**

**Minimal example**
<details>
<summary>Expand example Snakefile</summary>
<pre>
#!/usr/bin/env snakemake

localrules:
    test_rule

samples=list(x for x in range(0,1000))

rule test_rule:
  output:
    txt="output/{sample}.txt"
  shell:
    """echo SCIENCE > {output.txt} \
&&
sleep 10
"""

rule all:
  input:
    expand("output/{sample}.txt", sample=samples)
</pre>
</details>

<details>
<summary>Expand Profile</summary>
<pre>
executor: cluster-generic
cluster-generic-submit-cmd:
  mkdir -p results/logs/cluster/{rule}/ &&
  sbatch
    --parsable
    --cpus-per-task={threads}
    --time={resources.time}
    --mem={resources.mem_mb}
    --job-name=smk-{rule}
    --output=results/logs/cluster/{rule}/{jobid}.out
    --error=results/logs/cluster/{rule}/{jobid}.err
    --partition={resources.partition}
default-resources:
  - time=1440
  - partition='large'
  - tmpdir='/tmp'
local-cores: 4
jobs: 600
cores: 48  # Match maximum node size.
latency-wait: 120
keep-going: True
rerun-incomplete: True
printshellcmds: True
scheduler: greedy
use-conda: True
conda-frontend: mamba
cluster-generic-cancel-cmd: scancel
</pre>
</details>

<details>
<summary>Log of running the above Snakefile using cluster-generic</summary>
<pre>
Building DAG of jobs...
Using shell: /usr/bin/bash
Provided remote nodes: 600
Job stats:
job          count
---------  -------
all              1
test_rule     1000
total         1001

Select jobs to execute...
Execute 600 jobs...

[Mon May 27 10:56:12 2024]
localrule test_rule:
    output: output/141.txt
    jobid: 142
    reason: Missing output files: output/141.txt
    wildcards: sample=141
    resources: mem_mb=1000, mem_mib=954, disk_mb=1000, disk_mib=954, tmpdir=/tmp, time=1440, partition=large

echo SCIENCE > output/141.txt &&
sleep 10
...more job messages...
</pre>
</details>

**Additional context**

**Comment by cmeesters** [comment]
Does this happen too, if you use the `snakemake-executor-slurm`?

**Comment by tbigot** [comment]
It happens to me as well, using `snakemake-executor-slurm`.

**Comment by cmeesters** [comment]
Do your workflows make the distinction [between `localrules` and non-local rules (all others)](https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#local-rules)?

**Comment by tbigot** [comment]
Yes, they make this distinction using this formalism:

```py
rule some_local_rule:
    localrule: True
    input: "some_file.ext"
    …

rule some_dist_rule:
    input: "some_file_{x}.ext"
    …
```

**Comment by ypradat** [comment]
Hello, I am running into the same issue using snakemake `v.8.25.3` and `snakemake-executor-slurm` as well.

**Comment by VPestel** [comment]
Hello, I'm running into the same problem for a long time and I'm willing to solve it.

In the current state, I don't really understand what `local-cores` is supposed to do, but I haven't find any situation in which it seems to do something (correct me if I'm wrong).

Anyway, my understanding is that, when running with a non-local executors (e.g. slurm executor):
- `--cores` should be the maximum number of cores attributed to a job submitted
- `--jobs` should be the maximum number of jobs to be submitted
- `--local-cores` the number of cores to be used by the "main" process to execute localrules.

@cmeesters can you confirm this? 
If we converge on a clear definition of it, I can try to come up with an implementation.

**Comment by cmeesters** [comment]
yes, apart from the `--cores`, which I consider not useful due to uneven resources of submitted jobs: I rather rely on resource definitions and partition configurations as of the current implementation of the plugin. Does this help?

Sorry, I have been sick for a while, and I am trying to work myself through the issues, which piled up.
